onnxruntime/core/providers/coreml/builders/impl/split_op_builder.cc

```diff
@@ -0,0 +1,160 @@
+    auto split_sizes = split_span.size();
+    num_outputs = SafeInt<uint64_t>(split_sizes);
+    for (size_t i = 0; i < split_sizes; i++) {
+      coreml_splitnd->add_splitsizes(SafeInt<uint64_t>(split_span[i]));
+    }
+  } else if (node.SinceVersion() < 18) {
+    coreml_splitnd->set_numsplits(num_outputs);
+  } else {
+    num_outputs = SafeInt<uint64_t>(helper.Get("num_outputs", -1));
+    auto split_dim_size = data_shape[HandleNegativeAxis(axis, data_shape.size())];
+    uint64_t chunk_size = narrow<uint64_t>(std::ceil(float(split_dim_size) / num_outputs));
+    uint64_t remainder = split_dim_size % chunk_size;
+    if (remainder) {
+      // uneven
+      auto split_sizes = std::vector<uint64_t>(num_outputs, chunk_size);
+      split_sizes.back() = remainder;
+      for (size_t i = 0; i < split_sizes.size(); i++) {
+        coreml_splitnd->add_splitsizes(SafeInt<uint64_t>(split_sizes[i]));
+      }
+    } else {
+      // even
+      coreml_splitnd->set_numsplits(num_outputs);
+    }
+  }
+
+  *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
+  // variadic number of outputs. Calculated based on the length of the given splitSizes if provided.
+  // Otherwise, uses attribute value 'num_outputs'.
+  for (uint64_t i = 0; i < num_outputs; i++) {
+    *layer->mutable_output()->Add() = node.OutputDefs()[i]->Name();
+  }
+  model_builder.AddLayer(std::move(layer));
+
+  return Status::OK();
+}
+
+#endif
+
+// Operator support related
+
+bool SplitOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
+                                       const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+
+  NodeAttrHelper helper(node);
+  const auto axis = helper.Get("axis", 0);
+  const auto num_outputs = helper.Get("num_outputs", -1);
+
+  std::vector<int64_t> input_shape;
+  if (!GetStaticShape(*input_defs[0], input_shape, logger))
+    return false;
+
+  if (input_defs.size() > 1 && input_defs[1]->Exists()) {
+    if (!CheckIsConstantInitializer(*input_defs[1], input_params.graph_viewer, logger, "'split'")) {
+      return false;
+    }
+    const auto split_shape = *input_defs[1]->Shape();
+    if (split_shape.dim_size() < 2) {
+      LOGS(logger, VERBOSE) << "CoreML SplitND requires to produce at least 2 outputs.";
+      return false;
+    }
+  } else {
+    if (node.SinceVersion() >= 18) {
+      if (num_outputs < 2) {
+        LOGS(logger, VERBOSE) << "Invalid num_outputs. The value can not be lower than 1.\n"
+                              << "CoreML SplitND requires at least 2 outputs. num_outputs: " << num_outputs;
+        return false;
+      }
+      if (num_outputs != input_shape[HandleNegativeAxis(axis, input_shape.size())]) {
+        LOGS(logger, VERBOSE) << "Invalid num_outputs provided.\n."
+                              << "The value should equal to the size of dimension being split. num_outputs: "
+                              << num_outputs;
```

Comment:
Doesn't num_outputs have to equal node.OutputDefs().size() and just be >= the input shape for the axis? It doesn't have to evenly split that axis, and for each split to have a value of 1 for axis.

e.g. if axis==0, input shape is {4, 4} if num_outputs is 2, I would have expected 2 outputs of shape {2, 4}, but that fails the check that num_outputs is exactly equal to the input shape for the axis.

e.g. SplitOperatorTest.Split18_NumOutputs_EvenSplit and Split18_NumOutputs_UnevenSplit

Reply:
should be fixed now. this was wrong and was under my previous impression that the number_outputs was inferred from the size of split input provided..